### PR TITLE
QFM surface

### PR DESCRIPTION
**Pull request by ejpaul**
This pull request implements a quadratic-flux minimizing surface solver, defined as the minimizer of the objective function

f =  int d^2 x  (B \cdot n)^2/int d^2 x B^2

at fixed flux surface label (e.g., volume, area, or toroidal flux). This optimization problem can be solved using a penalty formulation with LBFGS or with the SLSQP constrained optimization algorithm. Much of the code was modeled after the Boozer surfaces.

I started this branch some time ago so I am not sure how seamless this merge will be. Let me know what changes I should make.

### CODE REVIEW

**Comment by florianwechsung** [comment]
Thank you for adding this @ejpaul ! This week is busy with Simons meetings, but I should have time next week to look at this in detail

**Comment by florianwechsung** [comment]
Hi @ejpaul, 

This looks pretty good already! Our automated testing on github has changed a fair bit since April, can you please merge master into your branch so that we get those changes?

I also noticed that you are computing `n = xphi \cross xtheta` (and it's derivative) by hand. That's not required, you can just call `surface.normal()`, `surface.dnormal_by_dcoeff()`, and `surface. d2normal_by_dcoeffdcoeff()`.

**Comment by florianwechsung** [comment]
Great, we're now getting the up-to-date testing system. There are a few tests and linting issues. Also, please move the example file into `examples/1_Simple` and add a line to `run_serial_examples` so that it's included in the tests.

**Comment by ejpaul** [comment]
There is still a failing test, but this appears unrelated to my contributions.

**Comment by florianwechsung** [comment]
Thanks! I've triggered a rerun, looks like some server timed out.

**Review by florianwechsung — APPROVED** [review]

**Comment by florianwechsung** [comment]
That seems to have done the trick. From my side this now looks good. Maybe @andrewgiuliani has more requests

**Review by andrewgiuliani — APPROVED** [review]
Looks good to me

File: src/simsopt/geo/qfmsurface.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        where f(x) is the QFM residual defined in surfaceobjectives.py. This is
+        done using LBFGS.
+        """
+
+        s = self.surface
+        x = s.get_dofs()
+        fun = lambda x: self.qfm_penalty_constraints(
+            x, derivatives=1, constraint_weight=constraint_weight)
+        res = minimize(
+            fun, x, jac=True, method='L-BFGS-B',
```
[the remark this answers is not in the document]
Comment (florianwechsung) [reply src/simsopt/geo/qfmsurface.py:133]: scipy doesn't have L-BFGS

**Comment by florianwechsung** [comment]
Thanks @ejpaul and apologies for the delay from our side. Merged!

**Pull request merged**